lib/Sema/CSFix.cpp

```diff
@@ -1148,6 +1148,10 @@ bool NotCompileTimeConst::diagnose(const Solution &solution, bool asNote) const
     if (isa<EnumElementDecl>(d)) {
       return true;
     }
+    // We should allow using a const variable reference
+    if (d->getAttrs().getAttribute(DeclAttrKind::DAK_CompileTimeConst)) {
+      return true;
```

Comment:
Why is diagnose returning true here without producing a diagnostic? Is the fix being recorded although it shouldn’t? The consequences of this could be a crash in SILGen.

Reply:
@nkcsgexi The check is being done too late. By the time you end up in CSFix the constraint system has already failed.

Reply:
This seems to be the kind of CS fix that won't fail the CS system, like how we generate warnings.

Reply:
```
NotCompileTimeConst::NotCompileTimeConst(ConstraintSystem &cs, Type paramTy,
                                         ConstraintLocator *locator):
  ContextualMismatch(cs, FixKind::NotCompileTimeConst, paramTy,
                     cs.getASTContext().TheEmptyTupleType, locator,
                     /*warning*/true) {}
```
This is specified via true in the warning field.

Reply:
That’s better but we shouldn’t record a fix if it’s not going to be diagnosed.

Reply:
hmm, I thought it's fine to record more than we need and later refine/diagnose as needed.

Reply:
You can do what other fixes do where they declare a static `attempt` and put all the conditions there, although it’s okay to record warning fixes but it’s not preferred.

Reply:
Can you decide if it’s an error at the point before the fix it’s recorded and only record it if the error condition holds? Also setting warning to true is confusing when the diagnostic is actually an error and not a warning

Reply:
Here is the code that records the fix in CSSimplify.cpp, matchCallArguments():

```
      if (!argument.isCompileTimeConst() && param.isCompileTimeConst()) {
        auto *locator = cs.getConstraintLocator(loc);
        SourceRange range;
        // simplify locator so the anchor is the exact argument.
        locator = simplifyLocator(cs, locator, range);
        if (locator->getPath().empty() &&
            locator->getAnchor().isExpr(ExprKind::UnresolvedMemberChainResult)) {
          locator =
            cs.getConstraintLocator(cast<UnresolvedMemberChainResultExpr>(
              locator->getAnchor().get<Expr*>())->getChainBase());
        }
        cs.recordFix(NotCompileTimeConst::create(cs, paramTy, locator));
      }
```

I think you should be able to move the validity condition to matchCallArguments() and check it in its entirely there. In fact it's not clear to me why argument.isCompileTimeConst() can be false if you're passing in a _const ParamDecl. Maybe there's another issue further upstream?

Reply:
Thank you! Will update the PR.